### ISSUE TITLE
go to definition only jump to correct location when specific file was opened once

### ISSUE THREAD

**Issue by tarik-jin** [clangd/clangd#1348]
when jumps to a fresh function (e.g., press F12 go to definition in vscode),  clangd only jumps to header file which contains declaration, c implementation file will be indexed only when I open the file manually, and after then, go to definition for this symbol works, 
does this mean the compile_commands.json file I used provide incomplete information when I indexed the whole project?
or maybe my index file in  .cache/clangd/index folder was outdated and need to be deleted and regenerated?

**Comment by tarik-jin** [comment]
I delete my local whole .cache/clangd folder and reindex wholly, it seems that even a new file was indexed showed in the output of vscode clangd message box, still can't go to definition unless open the specific manually

**Comment by HighCommander4** [comment]
Does the file containing the function's definition have an entry in the `compile_commands.json`?

**Comment by tarik-jin** [comment]
yes, and I can see the new generated *.idx file in the .cache/clangd/index folder
besides, seems go to definition works only after astworker building file *** which was shown in clangd output in vscode,   
do indexed *.cc and astworker building file *.cc do different things, right?

**Comment by HighCommander4** [comment]
> yes, and I can see the new generated *.idx file in the .cache/clangd/index folder 

Is it possible that the definition is conditional on an `#ifdef` and it's inactive when the file is indexed (e.g. the definition is inside an `#ifdef SOME_MACRO` block, and the file's command in `compile_commands.json` does not contain `-DSOME_MACRO`)?

> besides, seems go to definition works only after astworker building file *** which was shown in clangd output in vscode, do indexed *.cc and astworker building file *.cc do different things, right?

I assume you're referring here to the output showing "astworker building file" for the file in which you're **invoking** the go-to-definition.

If so, that's expected: the AST is needed to find the name of the symbol under the cursor, and then the index is used to find the definition of that symbol.

**Comment by tarik-jin** [comment]
>I assume you're referring here to the output showing "astworker building file" for the file in which you're invoking the go-to-definition.

yes
>If so, that's expected: the AST is needed to find the name of the symbol under the cursor, and then the index is used to find the definition of that symbol.

then the question is when the file astworker is called?
indexed *.cc was carried out for all files when I open a c file in a fresh project, while astworker was called only when I open specific file, Is this right?
 Is there any option or configuration that could call astworker for all files directly?

**Comment by HighCommander4** [comment]
> then the question is when the file astworker is called? indexed *.cc was carried out for all files when I open a c file in a fresh project, while astworker was called only when I open specific file, Is this right?

Yep.

> Is there any option or configuration that could call astworker for all files directly?

There is not; see some related discussion in #1165.

In any case, it seems like we're now talking about something different (the performance issue of how soon go-to-definition is available after opening a file) than the original issue description (which was about go-to-definition opening a header file instead of the source file containing the definition)?

**Comment by tarik-jin** [comment]
so, if we:
1. index a whole project from scratch, after finishing
2. open a *.c file, click a function call point
3. we are supposed to jump to function definition c file directly, right?
    we don't need to manual open the definition file to launch astworker to enable this jump, right?

**Comment by HighCommander4** [comment]
> so, if we:
> 
>     1. index a whole project from scratch, after finishing
> 
>     2. open a *.c file, click a function call point
> 
>     3. we are supposed to jump to function definition c file directly, right?

You should, yes.

>        we don't need to manual open the definition file to launch astworker to enable this jump, right?

That's right; the AST is only needed for the file containing the call, which is already open (and the AST build is triggered automatically when the file is opened).

**Comment by tarik-jin** [comment]
when click Go to definition, which file or something will reply the definition location?
Back to the issue, I checked that there is no conditional macro, and I wonder why when I open the c implementation file manually, Go to definition will work, _**does astworker overwrite something somewhere?**_, can I check this file to verify or figure out why I jump to declaration instead of definition firstly
since after I manually open the function definition file, all function in this definition file can be jumped correctly...

**Comment by tarik-jin** [comment]
I found a related issue https://github.com/clangd/clangd/issues/802, 
sorry for forgetting to mention that my project contains multiple CDB, but It seems not works for me by this way https://github.com/clangd/clangd/issues/802#issuecomment-864285877
Is there any standard configuration for projects contains multiple CDBs?

**Comment by tarik-jin** [comment]
this is the original way https://github.com/clangd/clangd/issues/1092#issuecomment-1087241220 I setup for my vscode workspace which results in this issues

**Comment by tarik-jin** [comment]
sorry for noise about this, I found it was due to mismatched version between clang used to generate CDB and clangd used to index. CDB was generated by clang 13.02, and index was carried out by clangd 14.0.6.
Change clangd from 14.0.6 to 13.0.0, Go-to-definition works well.
